{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "68795037-89f8-4059-a3e4-8a4c093eb31a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#You will need to be in the correct conda environment: env_camra\n",
    "import pandas as pd\n",
    "from openpyxl import Workbook,load_workbook\n",
    "import json\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "#from BeautifulSoup import BeautifulSoup as BS\n",
    "from bs4 import BeautifulSoup as BS\n",
    "import urllib3\n",
    "import requests\n",
    "import subprocess\n",
    "import json\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6118c988-41c9-482d-8289-43a8e970f38c",
   "metadata": {},
   "source": [
    "## Add your Assembly location here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1f642b93-b213-4400-b7d3-f2f990bf4e52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory and File good\n"
     ]
    }
   ],
   "source": [
    "BatchAssemblyLocation = \"/Users/dmatute/Documents/CAMRA/Bioinformatics/IFAIN/BatchB_250823/Assembly\"\n",
    "#BatchAssemblyLocation = \"/Users/dmatute/Documents/CAMRA/Bioinformatics/IFAIN/BatchA_121022/Assembly\"\n",
    "\n",
    "# Location of the sheet that will be populated\n",
    "excel_location= \"/Users/dmatute/Documents/CAMRA/Bioinformatics/Assembly_Metrics.xlsx\"\n",
    "\n",
    "if os.path.isdir(BatchAssemblyLocation) == True and os.path.isfile(excel_location) == True:\n",
    "    print(\"Directory and File good\")\n",
    "else:\n",
    "    print(\"BAD\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "72d19484-f9a6-4c9c-ac62-84095877f6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fix location if needed\n",
    "if BatchAssemblyLocation[-1] != \"/\": BatchAssemblyLocation = BatchAssemblyLocation + \"/\"\n",
    "\n",
    "# Get Batch Name\n",
    "batch_name = BatchAssemblyLocation.split(\"/\")[-3]\n",
    "\n",
    "# variable that checks if this batch has many checkm or a single one\n",
    "have_singlecheckm_file = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3610d141-4e4b-48d6-9205-da04bc18f27e",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "20e98bec-9506-4d81-821c-7097d027af70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to make line in Excel\n",
    "def append_to_excel(sample_name, sample_data, excel_location, message):\n",
    "    wb = load_workbook(filename=excel_location)\n",
    "    ws = wb['Assembly']\n",
    "    ws.append(sample_data)\n",
    "    wb.save(filename=excel_location)\n",
    "    print(message +\"    row added to \" + os.path.basename(excel_location))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "b8d134ec-9d7f-4f63-b5be-7038d6b3290e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function that gets quast for every sample\n",
    "def get_quast_info (quast_path):\n",
    "    with open(quast_path, 'r',) as file:\n",
    "        icount =0\n",
    "        quast_dict = {}\n",
    "        for line in file:\n",
    "            if icount >2:\n",
    "                value = line.split()[-1]\n",
    "                description = \" \".join(line.split()[:-1])\n",
    "                quast_dict[description] =value\n",
    "            icount += 1\n",
    "    return quast_dict\n",
    "\n",
    "## Finding MASH file, this file contains the mash output for all samples\n",
    "if os.path.isfile(BatchAssemblyLocation.replace(\"Assembly/\", \"QC/Mash/Mash_Taxa.txt\")):\n",
    "    mash_df = pd.read_csv(BatchAssemblyLocation.replace(\"Assembly/\", \"QC/Mash/Mash_Taxa.txt\"), sep=\",\", header=None)\n",
    "    mash_avaliable = True\n",
    "else:\n",
    "    print(batch_name+\" Mash_Taxa.txt unavailable\")\n",
    "    mash_avaliable = False\n",
    "\n",
    "\n",
    "## Finding Lab_assigned_Taxa file\n",
    "if os.path.isfile(BatchAssemblyLocation.replace(\"Assembly/\", \"\".join([\"Lab_Assigned_Taxa/\", batch_name,\".xlsx\"]))):\n",
    "    lab_taxa_loc = BatchAssemblyLocation.replace(\"Assembly/\", \"\".join([\"Lab_Assigned_Taxa/\", batch_name,\".xlsx\"]))\n",
    "    lab_taxa_df = pd.read_excel(lab_taxa_loc)\n",
    "    lab_taxa_df['sample_name'] =  lab_taxa_df['CAMRA ID'].str.cat(lab_taxa_df['S/N'].astype(str), sep='_S')\n",
    "    lab_taxa_available = True\n",
    "else:\n",
    "    print(batch_name+\" Lab_taxa file unavailable\")\n",
    "    lab_taxa_available = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "63103ac6-4a7d-4693-a5f6-e638c66a1fdd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">> NTC_SEQ_S96\n",
      "NTC_SEQ_S96    quast unavailable    Labtaxa xlsx unavailable    HTML unavailable    run.json unavailable    CheckM unavailable    Merqury unavailable    row added to Assembly_Metrics.xlsx\n",
      ">> CCI91_S11\n",
      "5509643.0 5682322.0\n",
      "CCI91_S11    row added to Assembly_Metrics.xlsx\n",
      ">> CCI105_S25\n",
      "5796485.0 5682322.0\n",
      "CCI105_S25    row added to Assembly_Metrics.xlsx\n",
      ">> CCI136_S56\n",
      "6003371.0 5682322.0\n",
      "CCI136_S56    row added to Assembly_Metrics.xlsx\n",
      ">> CCI152_S72\n",
      "5489201.0 5682322.0\n",
      "CCI152_S72    row added to Assembly_Metrics.xlsx\n",
      ">> CCI87_S7\n",
      "5619095.0 5682322.0\n",
      "CCI87_S7    row added to Assembly_Metrics.xlsx\n",
      ">> CCI137_S57\n",
      "5289872.0 5682322.0\n",
      "CCI137_S57    row added to Assembly_Metrics.xlsx\n",
      ">> CCI153_S73\n",
      "5693088.0 5682322.0\n",
      "CCI153_S73    row added to Assembly_Metrics.xlsx\n",
      ">> CCI90_S10\n",
      "5747089.0 5682322.0\n",
      "CCI90_S10    row added to Assembly_Metrics.xlsx\n",
      ">> CCI104_S24\n",
      "5701098.0 5682322.0\n",
      "CCI104_S24    row added to Assembly_Metrics.xlsx\n",
      ">> CCI171_S91\n",
      "CCI171_S91    Labtaxa xlsx available, but error finding reference size    row added to Assembly_Metrics.xlsx\n",
      ">> CCI155_S75\n",
      "5528373.0 5682322.0\n",
      "CCI155_S75    row added to Assembly_Metrics.xlsx\n",
      ">> CCI131_S51\n",
      "5466659.0 5682322.0\n",
      "CCI131_S51    row added to Assembly_Metrics.xlsx\n",
      ">> .DS_Store\n",
      ">> CCI168_S88\n",
      "5594395.0 5682322.0\n",
      "CCI168_S88    row added to Assembly_Metrics.xlsx\n",
      ">> CCI86_S6\n",
      "5526660.0 5682322.0\n",
      "CCI86_S6    row added to Assembly_Metrics.xlsx\n",
      ">> CCI102_S22\n",
      "2511694.0 5682322.0\n",
      "CCI102_S22    row added to Assembly_Metrics.xlsx\n",
      ">> CCI128_S48\n",
      "3025672.0 5682322.0\n",
      "CCI128_S48    row added to Assembly_Metrics.xlsx\n",
      ">> CCI96_S16\n",
      "5374365.0 5682322.0\n",
      "CCI96_S16    row added to Assembly_Metrics.xlsx\n",
      ">> CCI169_S89\n",
      "5214823.0 5682322.0\n",
      "CCI169_S89    row added to Assembly_Metrics.xlsx\n",
      ">> CCI103_S23\n",
      "5251060.0 5682322.0\n",
      "CCI103_S23    row added to Assembly_Metrics.xlsx\n",
      ">> CCI129_S49\n",
      "5368033.0 5682322.0\n",
      "CCI129_S49    row added to Assembly_Metrics.xlsx\n",
      ">> CCI97_S17\n",
      "5553249.0 5682322.0\n",
      "CCI97_S17    row added to Assembly_Metrics.xlsx\n",
      ">> CCI154_S74\n",
      "5697318.0 5682322.0\n",
      "CCI154_S74    row added to Assembly_Metrics.xlsx\n",
      ">> CCI170_S90\n",
      "5362379.0 5682322.0\n",
      "CCI170_S90    row added to Assembly_Metrics.xlsx\n",
      ">> CCI130_S50\n",
      "5438199.0 5682322.0\n",
      "CCI130_S50    row added to Assembly_Metrics.xlsx\n",
      ">> CCI164_S84\n",
      "5487389.0 5682322.0\n",
      "CCI164_S84    row added to Assembly_Metrics.xlsx\n",
      ">> CCI85_S5\n",
      "5524283.0 5682322.0\n",
      "CCI85_S5    row added to Assembly_Metrics.xlsx\n",
      ">> CCI140_S60\n",
      "5698937.0 5682322.0\n",
      "CCI140_S60    row added to Assembly_Metrics.xlsx\n",
      ">> CCI124_S44\n",
      "5441028.0 5682322.0\n",
      "CCI124_S44    row added to Assembly_Metrics.xlsx\n",
      ">> CCI159_S79\n",
      "5116063.0 5682322.0\n",
      "CCI159_S79    row added to Assembly_Metrics.xlsx\n",
      ">> CCI117_S37\n",
      "5565686.0 5682322.0\n",
      "CCI117_S37    row added to Assembly_Metrics.xlsx\n",
      ">> CCI158_S78\n",
      "5553546.0 5682322.0\n",
      "CCI158_S78    row added to Assembly_Metrics.xlsx\n",
      ">> CCI116_S36\n",
      "5125406.0 5682322.0\n",
      "CCI116_S36    row added to Assembly_Metrics.xlsx\n",
      ">> CCI141_S61\n",
      "5088901.0 5682322.0\n",
      "CCI141_S61    row added to Assembly_Metrics.xlsx\n",
      ">> CCI165_S85\n",
      "5546688.0 5682322.0\n",
      "CCI165_S85    row added to Assembly_Metrics.xlsx\n",
      ">> CCI125_S45\n",
      "5352884.0 5682322.0\n",
      "CCI125_S45    row added to Assembly_Metrics.xlsx\n",
      ">> CCI89_S9\n",
      "5756198.0 5682322.0\n",
      "CCI89_S9    row added to Assembly_Metrics.xlsx\n",
      ">> CCI84_S4\n",
      "5621451.0 5682322.0\n",
      "CCI84_S4    row added to Assembly_Metrics.xlsx\n",
      ">> CCI110_S30\n",
      "5589167.0 5682322.0\n",
      "CCI110_S30    row added to Assembly_Metrics.xlsx\n",
      ">> CCI123_S43\n",
      "5350517.0 5682322.0\n",
      "CCI123_S43    row added to Assembly_Metrics.xlsx\n",
      ">> CCI147_S67\n",
      "5454510.0 5682322.0\n",
      "CCI147_S67    row added to Assembly_Metrics.xlsx\n",
      ">> CCI163_S83\n",
      "5206344.0 5682322.0\n",
      "CCI163_S83    row added to Assembly_Metrics.xlsx\n",
      ">> CCI109_S29\n",
      "5641982.0 5682322.0\n",
      "CCI109_S29    row added to Assembly_Metrics.xlsx\n",
      ">> CCI122_S42\n",
      "5401783.0 5682322.0\n",
      "CCI122_S42    row added to Assembly_Metrics.xlsx\n",
      ">> CCI162_S82\n",
      "5344419.0 5682322.0\n",
      "CCI162_S82    row added to Assembly_Metrics.xlsx\n",
      ">> CCI146_S66\n",
      "5279501.0 5682322.0\n",
      "CCI146_S66    row added to Assembly_Metrics.xlsx\n",
      ">> CCI108_S28\n",
      "5660557.0 5682322.0\n",
      "CCI108_S28    row added to Assembly_Metrics.xlsx\n",
      ">> CCI88_S8\n",
      "5752990.0 5682322.0\n",
      "CCI88_S8    row added to Assembly_Metrics.xlsx\n",
      ">> CCI111_S31\n",
      "5153615.0 5682322.0\n",
      "CCI111_S31    row added to Assembly_Metrics.xlsx\n",
      ">> CCI93_S13\n",
      "2277334.0 5682322.0\n",
      "CCI93_S13    row added to Assembly_Metrics.xlsx\n",
      ">> CCI107_S27\n",
      "5793313.0 5682322.0\n",
      "CCI107_S27    row added to Assembly_Metrics.xlsx\n",
      ">> CCI149_S69\n",
      "5410873.0 5682322.0\n",
      "CCI149_S69    row added to Assembly_Metrics.xlsx\n",
      ">> CCI134_S54\n",
      "5266095.0 5682322.0\n",
      "CCI134_S54    row added to Assembly_Metrics.xlsx\n",
      ">> CCI150_S70\n",
      "5510446.0 5682322.0\n",
      "CCI150_S70    row added to Assembly_Metrics.xlsx\n",
      ">> CCI135_S55\n",
      "5186463.0 5682322.0\n",
      "CCI135_S55    row added to Assembly_Metrics.xlsx\n",
      ">> CCI151_S71\n",
      "5704557.0 5682322.0\n",
      "CCI151_S71    row added to Assembly_Metrics.xlsx\n",
      ">> CCI83_S3\n",
      "5488083.0 5682322.0\n",
      "CCI83_S3    row added to Assembly_Metrics.xlsx\n",
      ">> CCI92_S12\n",
      "5745028.0 5682322.0\n",
      "CCI92_S12    row added to Assembly_Metrics.xlsx\n",
      ">> CCI106_S26\n",
      "5796937.0 5682322.0\n",
      "CCI106_S26    row added to Assembly_Metrics.xlsx\n",
      ">> CCI148_S68\n",
      "5542015.0 5682322.0\n",
      "CCI148_S68    row added to Assembly_Metrics.xlsx\n",
      ">> CCI119_S39\n",
      "5507463.0 5682322.0\n",
      "CCI119_S39    row added to Assembly_Metrics.xlsx\n",
      ">> CCI157_S77\n",
      "5520772.0 5682322.0\n",
      "CCI157_S77    row added to Assembly_Metrics.xlsx\n",
      ">> CCI173_S93\n",
      "5462434.0 5682322.0\n",
      "CCI173_S93    row added to Assembly_Metrics.xlsx\n",
      ">> CCI133_S53\n",
      "5486671.0 5682322.0\n",
      "CCI133_S53    row added to Assembly_Metrics.xlsx\n",
      ">> CCI100_S20\n",
      "5592122.0 5682322.0\n",
      "CCI100_S20    row added to Assembly_Metrics.xlsx\n",
      ">> CCI94_S14\n",
      "5440617.0 5682322.0\n",
      "CCI94_S14    row added to Assembly_Metrics.xlsx\n",
      ">> CCI82_S2\n",
      "5765451.0 5682322.0\n",
      "CCI82_S2    row added to Assembly_Metrics.xlsx\n",
      ">> CCI101_S21\n",
      "5452546.0 5682322.0\n",
      "CCI101_S21    row added to Assembly_Metrics.xlsx\n",
      ">> CCI95_S15\n",
      "5454620.0 5682322.0\n",
      "CCI95_S15    row added to Assembly_Metrics.xlsx\n",
      ">> CCI118_S38\n",
      "5450691.0 5682322.0\n",
      "CCI118_S38    row added to Assembly_Metrics.xlsx\n",
      ">> CCI172_S92\n",
      "5461443.0 5682322.0\n",
      "CCI172_S92    row added to Assembly_Metrics.xlsx\n",
      ">> CCI156_S76\n",
      "5459800.0 5682322.0\n",
      "CCI156_S76    row added to Assembly_Metrics.xlsx\n",
      ">> CCI132_S52\n",
      "5120570.0 5682322.0\n",
      "CCI132_S52    row added to Assembly_Metrics.xlsx\n",
      ">> CCI142_S62\n",
      "5460019.0 5682322.0\n",
      "CCI142_S62    row added to Assembly_Metrics.xlsx\n",
      ">> CCI166_S86\n",
      "5472695.0 5682322.0\n",
      "CCI166_S86    row added to Assembly_Metrics.xlsx\n",
      ">> CCI98_S18\n",
      "5431474.0 5682322.0\n",
      "CCI98_S18    row added to Assembly_Metrics.xlsx\n",
      ">> CCI126_S46\n",
      "5364942.0 5682322.0\n",
      "CCI126_S46    row added to Assembly_Metrics.xlsx\n",
      ">> CCI115_S35\n",
      "5364746.0 5682322.0\n",
      "CCI115_S35    row added to Assembly_Metrics.xlsx\n",
      ">> CCI114_S34\n",
      "5360310.0 5682322.0\n",
      "CCI114_S34    row added to Assembly_Metrics.xlsx\n",
      ">> CCI81_S1\n",
      "5473177.0 5682322.0\n",
      "CCI81_S1    row added to Assembly_Metrics.xlsx\n",
      ">> CCI167_S87\n",
      "5322903.0 5682322.0\n",
      "CCI167_S87    row added to Assembly_Metrics.xlsx\n",
      ">> CCI143_S63\n",
      "5506821.0 5682322.0\n",
      "CCI143_S63    row added to Assembly_Metrics.xlsx\n",
      ">> CCI99_S19\n",
      "5582508.0 5682322.0\n",
      "CCI99_S19    row added to Assembly_Metrics.xlsx\n",
      ">> CCI127_S47\n",
      "5265173.0 5682322.0\n",
      "CCI127_S47    row added to Assembly_Metrics.xlsx\n",
      ">> CCI138_S58\n",
      "21720.0 5682322.0\n",
      "CCI138_S58    row added to Assembly_Metrics.xlsx\n",
      ">> CCI112_S32\n",
      "5623239.0 5682322.0\n",
      "CCI112_S32    row added to Assembly_Metrics.xlsx\n",
      ">> CCI121_S41\n",
      "5368235.0 5682322.0\n",
      "CCI121_S41    row added to Assembly_Metrics.xlsx\n",
      ">> CCI161_S81\n",
      "5635879.0 5682322.0\n",
      "CCI161_S81    row added to Assembly_Metrics.xlsx\n",
      ">> CCI145_S65\n",
      "5559956.0 5682322.0\n",
      "CCI145_S65    row added to Assembly_Metrics.xlsx\n",
      ">> CCI120_S40\n",
      "5680948.0 5682322.0\n",
      "CCI120_S40    row added to Assembly_Metrics.xlsx\n",
      ">> NTC_IFAIN1_S94\n",
      "NTC_IFAIN1_S94    quast unavailable    Labtaxa xlsx unavailable    HTML unavailable    run.json unavailable    CheckM unavailable    Merqury unavailable    row added to Assembly_Metrics.xlsx\n",
      ">> NTC_IFAIN2_S95\n",
      "NTC_IFAIN2_S95    quast unavailable    Labtaxa xlsx unavailable    HTML unavailable    run.json unavailable    CheckM unavailable    Merqury unavailable    row added to Assembly_Metrics.xlsx\n",
      ">> CCI144_S64\n",
      "5579458.0 5682322.0\n",
      "CCI144_S64    row added to Assembly_Metrics.xlsx\n",
      ">> CCI160_S80\n",
      "4835766.0 5682322.0\n",
      "CCI160_S80    row added to Assembly_Metrics.xlsx\n",
      ">> CCI139_S59\n",
      "5310166.0 5682322.0\n",
      "CCI139_S59    row added to Assembly_Metrics.xlsx\n",
      ">> CCI113_S33\n",
      "5383730.0 5682322.0\n",
      "CCI113_S33    row added to Assembly_Metrics.xlsx\n"
     ]
    }
   ],
   "source": [
    "#ITERATE the SAMPLES\n",
    "have_singlecheckm_file = False\n",
    "empty = \"\"\n",
    "for sample_name in os.listdir(BatchAssemblyLocation):\n",
    "    print(\">>\", sample_name)\n",
    "    if sample_name[0] != \".\" :\n",
    "\n",
    "        message = sample_name\n",
    "        sample_path = BatchAssemblyLocation + sample_name\n",
    "\n",
    "        ## QUAST\n",
    "        if os.path.isfile(sample_path + \"/details/\" + sample_name+ \"_quast_report.txt\" ):\n",
    "            quast_loc = sample_path + \"/details/\" + sample_name+ \"_quast_report.txt\" \n",
    "            quast_dict = get_quast_info(quast_loc)\n",
    "            \n",
    "            quast_longest_contg = float(quast_dict['Largest contig'])\n",
    "            quast_N50 = float(quast_dict['N50'])\n",
    "            quast_N90 = float(quast_dict['N90'])\n",
    "            quast_L50 = float(quast_dict['L50'])\n",
    "            quast_L90 = float(quast_dict['L90'])\n",
    "            quast_ASM_SIZE = float(quast_dict['Total length'])\n",
    "        else:\n",
    "            message = message +\"    quast unavailable\"\n",
    "            quast_longest_contg = empty\n",
    "            quast_N50 = empty\n",
    "            quast_N90 = empty\n",
    "            quast_L50 = empty\n",
    "            quast_L90 = empty\n",
    "            quast_ASM_SIZE = empty     \n",
    "            \n",
    "\n",
    "        #LAB TAXA\n",
    "        lab_genusspecies = '' #made for \"#Lab expected genome size \" validation\n",
    "        if lab_taxa_available == True:\n",
    "            find_name = sample_name.split('_')[0]\n",
    "            lab_output = lab_taxa_df.loc[lab_taxa_df['CAMRA ID'] == find_name].values. flatten().tolist()\n",
    "            if lab_output!=[]:\n",
    "                #seperate genus and species form subspecies \n",
    "                lab_taxasplit = lab_output[3].split(\" \")\n",
    "                lab_genusspecies = \" \".join(lab_taxasplit[:2])\n",
    "                lab_subspecies = \" \".join(lab_taxasplit[2:])\n",
    "                \n",
    "                try:\n",
    "                    ref_command = [\"datasets\", \"summary\", \"genome\", \"taxon\", lab_genusspecies ,\"--reference\"]\n",
    "                    ref_summary = subprocess.run(ref_command, capture_output=True, text=True)\n",
    "                    ref_summary = ref_summary.stdout.splitlines()\n",
    "                    ref_summary_dict = json.loads(ref_summary[0])\n",
    "                    lab_length = float(ref_summary_dict[\"reports\"][0][\"assembly_stats\"]['total_sequence_length'])\n",
    "                    asmsize_div_labsize = round(quast_ASM_SIZE/float(lab_length),4)*100\n",
    "                except:\n",
    "                    message = message + \"    Labtaxa xlsx available, but error finding reference size\"\n",
    "                    asmsize_div_labsize = empty\n",
    "            else:\n",
    "                message = message +\"    Labtaxa xlsx unavailable\"\n",
    "                lab_genusspecies = empty\n",
    "                lab_subspecies = empty\n",
    "                lab_length = empty\n",
    "                asmsize_div_labsize = empty\n",
    "            \n",
    "        else:\n",
    "            message = message +\"    Labtaxa xlsx unavailable\"\n",
    "            lab_genusspecies = empty\n",
    "            lab_subspecies = empty\n",
    "            lab_length = empty\n",
    "            asmsize_div_labsize = empty\n",
    "\n",
    "\n",
    "        \n",
    "        ## HTML \n",
    "        if os.path.isfile(sample_path + \"/\" + \"AssemblyReport.html\"):\n",
    "            HTML_loc = sample_path + \"/\" + \"AssemblyReport.html\"\n",
    "            with open(HTML_loc, 'r', encoding='utf-8') as file:\n",
    "                # Read the contents of the file\n",
    "                html_content = file.read()\n",
    "                # Parse the HTML using BeautifulSoup\n",
    "                soup = BS(html_content, 'html.parser')\n",
    "                div = soup.find_all('div')[-1]\n",
    "                html_av_read_size = float(div.find_all('td')[-1].text)\n",
    "                html_numb_reads = float(div.find_all('td') [-2].text)\n",
    "        else:\n",
    "            message = message +\"    HTML unavailable\"\n",
    "            html_numb_reads = empty\n",
    "            html_av_read_size = empty\n",
    "\n",
    "\n",
    "        ## JSON with ASM output metrics\n",
    "        if os.path.isfile(sample_path + \"/details/\" + sample_name+ \"_run_details.json\" ):\n",
    "            run_loc = sample_path + \"/details/\" + sample_name+ \"_run_details.json\" \n",
    "            with open(run_loc, 'r') as json_file:\n",
    "                run_json = json.load(json_file)\n",
    "                json_read_depth = round(float(run_json['contig_filtering']['average depth (short reads)']),1)\n",
    "                json_above_thrshld = float(run_json['contig_filtering']['num contigs above thresholds'])\n",
    "                json_below_thrshld = float(run_json['contig_filtering']['num contigs below thresholds'])\n",
    "                json_total_size = float(run_json['assembly']['contigs.fasta file size'])\n",
    "        else:\n",
    "            message = message +\"    run.json unavailable\"\n",
    "            json_read_depth = empty\n",
    "            json_above_thrshld = empty\n",
    "            json_below_thrshld = empty\n",
    "            json_total_size = empty\n",
    "\n",
    "        \n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "        #MASH\n",
    "        if mash_avaliable == True:\n",
    "            mash_output = mash_df.loc[mash_df[0] == sample_name].values.flatten().tolist()\n",
    "            if mash_output!=[]:\n",
    "                #seperate genus and species form subspecies \n",
    "                mash_ani = (1-float(mash_output[4]))*100\n",
    "                mash_taxasplit = mash_output[3].split(\" \")\n",
    "                mash_taxa_id = mash_output[2]\n",
    "                if mash_ani >= 97.5:\n",
    "                    mash_genusspecies = \" \".join(mash_taxasplit[:2])\n",
    "                    mash_subspecies = \" \".join(mash_taxasplit[2:])\n",
    "                elif mash_ani < 97.5 and mash_ani >=94 : \n",
    "                    mash_genusspecies = \" \".join(mash_taxasplit[:2])\n",
    "                    mash_subspecies = empty\n",
    "                elif mash_ani < 94 and mash_ani >= 80 :\n",
    "                    mash_genusspecies = \" \".join(mash_taxasplit[:1])\n",
    "                    mash_subspecies = empty\n",
    "                elif mash_ani < 80:\n",
    "                    mash_genusspecies = empty\n",
    "                    mash_subspecies = empty \n",
    "                mash_asm_id = mash_output[1]\n",
    "            else:\n",
    "                mash_asm_id = empty\n",
    "                mash_genusspecies = empty\n",
    "                mash_subspecies = empty\n",
    "                mash_ani = empty\n",
    "                mash_taxa_id = empty\n",
    "        else:\n",
    "            message = message +\"    Labtaxa xlsx unavailable\"\n",
    "            mash_asm_id = empty\n",
    "            mash_genusspecies = empty\n",
    "            mash_subspecies = empty\n",
    "            mash_ani = empty\n",
    "            mash_taxa_id = empty\n",
    "\n",
    "              \n",
    "\n",
    "        #CheckM\n",
    "        def checkm_mapping (checkm_dic,sample_name):\n",
    "            checkm_markerset = checkm_dic[sample_name][2]\n",
    "            checkm_completness = float(checkm_dic[sample_name][13])\n",
    "            checkm_contamination = float(checkm_dic[sample_name][14])\n",
    "            checkm_heterogeneity = float(checkm_dic[sample_name][15])\n",
    "            return checkm_markerset, checkm_completness, checkm_contamination, checkm_heterogeneity\n",
    "            \n",
    "        ## Finding CheckM file, each sample has their own checkm output\n",
    "        if os.path.isfile(BatchAssemblyLocation.replace(\"Assembly/\", f\"QC/CheckM/{sample_name}/checkm_quality_assessment.txt\")):\n",
    "            checkm_file = open(BatchAssemblyLocation.replace(\"Assembly/\", f\"QC/CheckM/{sample_name}/checkm_quality_assessment.txt\"), \"r\")\n",
    "            content=checkm_file.readlines()\n",
    "            for i in range (0,len(content)):\n",
    "                if content[i][0:len(sample_name)+2].strip() == sample_name:\n",
    "                    checkm_metrics = content[i].split()\n",
    "                    checkm_markerset = checkm_metrics[1]\n",
    "                    checkm_completness = float(checkm_metrics[12])\n",
    "                    checkm_contamination = float(checkm_metrics[13])\n",
    "                    checkm_heterogeneity = float(checkm_metrics[14])\n",
    "            checkm_file.close()\n",
    "        #Find checkM file when all the samples in the batch share a single output file\n",
    "        elif os.path.isfile(BatchAssemblyLocation.replace(\"Assembly/\", f\"QC/checkm_quality_assessment.txt\")):\n",
    "            if have_singlecheckm_file == False:\n",
    "                checkm_file = open(BatchAssemblyLocation.replace(\"Assembly/\", f\"QC/checkm_quality_assessment.txt\"), \"r\")\n",
    "                content=checkm_file.readlines()\n",
    "                checkm_dic = {}\n",
    "                for i in range (0,len(content)):\n",
    "                    if content[i][0:3].strip() == \"CCI\":\n",
    "                        checkm_metrics = content[i].split()\n",
    "                        checkm_sample_name = checkm_metrics[0].split(\"/\")[0]\n",
    "                        checkm_dic[checkm_sample_name] = checkm_metrics\n",
    "                have_singlecheckm_file = True\n",
    "                checkm_file.close()\n",
    "                checkm_merkerser, checkm_completness, checkm_contamination, checkm_heterogeneity = checkm_mapping(checkm_dic,sample_name)\n",
    "            else:\n",
    "                checkm_merkerser, checkm_completness, checkm_contamination, checkm_heterogeneity = checkm_mapping(checkm_dic,sample_name)\n",
    "        else:\n",
    "            checkm_markerset = empty\n",
    "            checkm_completness = empty\n",
    "            checkm_contamination = empty\n",
    "            checkm_heterogeneity = empty\n",
    "            message = message +\"    CheckM unavailable\"\n",
    "            \n",
    "\n",
    "        #Merqury\n",
    "        merqury_path = BatchAssemblyLocation.replace(\"Assembly/\", f\"QC/Merqury/{sample_name}\")\n",
    "        if os.path.isfile(f\"{merqury_path}/{sample_name}.qv\") and os.path.isfile(f\"{merqury_path}/{sample_name}.completeness.stats\") :\n",
    "            merqury_qv_file = open(f\"{merqury_path}/{sample_name}.qv\", \"r\")\n",
    "            merqury_com_file = open(f\"{merqury_path}/{sample_name}.completeness.stats\",\"r\")\n",
    "            content_qv=merqury_qv_file.readlines()\n",
    "            content_com=merqury_com_file.readlines()\n",
    "            merqury_qv = content_qv[0].split()[3]\n",
    "            if merqury_qv == \"inf\":\n",
    "                merqury_qv = 1000\n",
    "            else:\n",
    "                mercury_qv = float(merqury_qv)\n",
    "            merqury_completeness = float(content_com[0].split()[4])\n",
    "            merqury_qv_file.close()\n",
    "            merqury_com_file.close()\n",
    "        else:\n",
    "            message = message +\"    Merqury unavailable\"\n",
    "            merqury_qv = empty\n",
    "            merqury_completeness = empty\n",
    "\n",
    "        \n",
    "        #Calculations\n",
    "        if quast_ASM_SIZE != empty and json_total_size!=empty:\n",
    "            quastASM_div_jsonSize = round(quast_ASM_SIZE/json_total_size,4)*100\n",
    "        else: quastASM_div_jsonSize = empty\n",
    "            \n",
    "        if quast_ASM_SIZE != empty and quast_N90 != empty:\n",
    "            quastN90_div_quastASM = round(float(quast_dict['N90'])/quast_ASM_SIZE,4)*100\n",
    "        else:\n",
    "            quastN90_div_quastASM = empty\n",
    "        \n",
    "        \n",
    "        #MAKE EXCEL ROW\n",
    "        #if os.path.isfile(sample_path + \"/details/\" + sample_name+ \"_run_details.json\" ) and os.path.isfile(sample_path + \"/details/\" + sample_name+ \"_quast_report.txt\" ):\n",
    "        # print(sample_path)\n",
    "        if os.path.isdir(sample_path) :\n",
    " \n",
    "            sample_data = [sample_name, \n",
    "                           batch_name,\n",
    "                           lab_genusspecies,\n",
    "                           lab_subspecies, \n",
    "                           lab_length,\n",
    "                           html_numb_reads,\n",
    "                           html_av_read_size,\n",
    "                           json_read_depth, \n",
    "                           json_above_thrshld,\n",
    "                           json_below_thrshld, \n",
    "                           quast_longest_contg,\n",
    "                           quast_N50,\n",
    "                           quast_N90,\n",
    "                           quast_L50,\n",
    "                           quast_L90,\n",
    "                           quast_ASM_SIZE,\n",
    "                           json_total_size,\n",
    "                           quastASM_div_jsonSize,\n",
    "                           asmsize_div_labsize, \n",
    "                           quastN90_div_quastASM,\n",
    "                           mash_taxa_id,\n",
    "                           mash_asm_id,\n",
    "                           mash_genusspecies,\n",
    "                           mash_subspecies,\n",
    "                           mash_ani,\n",
    "                           checkm_markerset,\n",
    "                           checkm_completness,\n",
    "                           merqury_completeness,\n",
    "                           checkm_contamination,\n",
    "                           checkm_heterogeneity,\n",
    "                           merqury_qv]\n",
    "        #print(sample_data)\n",
    "\n",
    "\n",
    "\n",
    "        # SEND TO EXCEL\n",
    "        append_to_excel(sample_name, sample_data, excel_location, message)  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38063bba-55f9-4e14-a9dd-25e5aeb03406",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8919362-1571-4111-921c-57eea23e219b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#must be within these vlaues\n",
    "qc_metric = {\"quastASM_div_jsonSize\":95,\n",
    "             \"asmsize_div_labsize\":[195,95], \n",
    "\n",
    "# def pass_or_fail (sample_data)\n",
    "#     status = \"fail\"\n",
    "#     msg = \"\"\n",
    "#     if sample_data[15] == \"\":\n",
    "#         msg = \"Assembly Size = 0\"\n",
    "#         return status\n",
    "#     else:\n",
    "#         if sample_data[9]/sample_data[8]*100 > 74:\n",
    "#             msg = \"Bad reads are 74 of the total reas\n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6055a0af-416c-4517-82b5-c503b7e689ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCI10 ['CCI10/checkm_quality_assessment.txt:', 'CCI10_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1347', '10', '0', '0', '0', '99.70', '1.26', '20.00']\n",
      "CCI11 ['CCI11/checkm_quality_assessment.txt:', 'CCI11_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1347', '10', '0', '0', '0', '99.70', '1.26', '20.00']\n",
      "CCI12 ['CCI12/checkm_quality_assessment.txt:', 'CCI12_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '68', '1279', '11', '1', '0', '0', '90.41', '0.79', '7.14']\n",
      "CCI13 ['CCI13/checkm_quality_assessment.txt:', 'CCI13_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '108', '1214', '36', '1', '0', '0', '87.16', '2.46', '0.00']\n",
      "CCI14 ['CCI14/checkm_quality_assessment.txt:', 'CCI14_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1354', '4', '0', '0', '0', '99.99', '0.40', '0.00']\n",
      "CCI15 ['CCI15/checkm_quality_assessment.txt:', 'CCI15_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '97', '1233', '28', '1', '0', '0', '87.74', '2.00', '0.00']\n",
      "CCI16 ['CCI16/checkm_quality_assessment.txt:', 'CCI16_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '108', '1212', '38', '1', '0', '0', '87.16', '2.57', '0.00']\n",
      "CCI17 ['CCI17/checkm_quality_assessment.txt:', 'CCI17_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1353', '4', '0', '0', '0', '99.94', '0.52', '25.00']\n",
      "CCI18 ['CCI18/checkm_quality_assessment.txt:', 'CCI18_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '682', '655', '22', '0', '0', '0', '36.28', '1.77', '0.00']\n",
      "CCI19 ['CCI19/checkm_quality_assessment.txt:', 'CCI19_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '682', '655', '21', '1', '0', '0', '36.28', '1.84', '0.00']\n",
      "CCI1 ['CCI1/checkm_quality_assessment.txt:', 'CCI1_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1359', '0', '0', '0', '0', '0', '0.00', '0.00', '0.00']\n",
      "CCI20 ['CCI20/checkm_quality_assessment.txt:', 'CCI20_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '682', '655', '21', '1', '0', '0', '36.28', '1.84', '0.00']\n",
      "CCI21 ['CCI21/checkm_quality_assessment.txt:', 'CCI21_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '685', '654', '20', '0', '0', '0', '36.19', '1.60', '0.00']\n",
      "CCI22 ['CCI22/checkm_quality_assessment.txt:', 'CCI22_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '685', '654', '19', '1', '0', '0', '36.19', '1.67', '0.00']\n",
      "CCI23 ['CCI23/checkm_quality_assessment.txt:', 'CCI23_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '126', '1209', '20', '2', '2', '0', '86.12', '2.50', '0.00']\n",
      "CCI24 ['CCI24/checkm_quality_assessment.txt:', 'CCI24_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '124', '1210', '21', '3', '1', '0', '86.46', '2.53', '0.00']\n",
      "CCI25 ['CCI25/checkm_quality_assessment.txt:', 'CCI25_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '123', '1209', '23', '3', '1', '0', '86.47', '2.91', '0.00']\n",
      "CCI26 ['CCI26/checkm_quality_assessment.txt:', 'CCI26_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '123', '1211', '21', '3', '1', '0', '86.55', '2.79', '0.00']\n",
      "CCI27 ['CCI27/checkm_quality_assessment.txt:', 'CCI27_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1205', '22', '4', '1', '0', '85.97', '2.70', '0.00']\n",
      "CCI28 ['CCI28/checkm_quality_assessment.txt:', 'CCI28_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '124', '1208', '24', '2', '1', '0', '86.46', '2.81', '0.00']\n",
      "CCI29 ['CCI29/checkm_quality_assessment.txt:', 'CCI29_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '8', '1349', '2', '0', '0', '0', '99.13', '0.31', '0.00']\n",
      "CCI2 ['CCI2/checkm_quality_assessment.txt:', 'CCI2_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1353', '4', '0', '0', '0', '99.70', '0.44', '0.00']\n",
      "CCI30 ['CCI30/checkm_quality_assessment.txt:', 'CCI30_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1349', '8', '0', '0', '0', '99.70', '0.51', '0.00']\n",
      "CCI31 ['CCI31/checkm_quality_assessment.txt:', 'CCI31_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '4', '1350', '5', '0', '0', '0', '99.41', '0.47', '0.00']\n",
      "CCI32 ['CCI32/checkm_quality_assessment.txt:', 'CCI32_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '5', '1349', '5', '0', '0', '0', '99.40', '0.47', '0.00']\n",
      "CCI33 ['CCI33/checkm_quality_assessment.txt:', 'CCI33_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1349', '9', '0', '0', '0', '99.99', '0.55', '0.00']\n",
      "CCI34 ['CCI34/checkm_quality_assessment.txt:', 'CCI34_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1349', '9', '0', '0', '0', '99.99', '0.55', '0.00']\n",
      "CCI35 ['CCI35/checkm_quality_assessment.txt:', 'CCI35_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1351', '7', '0', '0', '0', '99.99', '1.00', '0.00']\n",
      "CCI36 ['CCI36/checkm_quality_assessment.txt:', 'CCI36_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1355', '3', '0', '0', '0', '99.99', '0.32', '0.00']\n",
      "CCI37 ['CCI37/checkm_quality_assessment.txt:', 'CCI37_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1355', '3', '0', '0', '0', '99.99', '0.32', '0.00']\n",
      "CCI38 ['CCI38/checkm_quality_assessment.txt:', 'CCI38_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '126', '1209', '20', '2', '2', '0', '86.12', '2.50', '0.00']\n",
      "CCI39 ['CCI39/checkm_quality_assessment.txt:', 'CCI39_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1208', '20', '2', '2', '0', '86.06', '2.50', '0.00']\n",
      "CCI3 ['CCI3/checkm_quality_assessment.txt:', 'CCI3_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1355', '3', '0', '0', '0', '99.99', '0.59', '0.00']\n",
      "CCI40 ['CCI40/checkm_quality_assessment.txt:', 'CCI40_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1207', '21', '2', '2', '0', '86.06', '2.60', '0.00']\n",
      "CCI41 ['CCI41/checkm_quality_assessment.txt:', 'CCI41_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1207', '21', '2', '2', '0', '86.06', '2.60', '0.00']\n",
      "CCI42 ['CCI42/checkm_quality_assessment.txt:', 'CCI42_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '126', '1208', '21', '2', '2', '0', '86.12', '2.60', '0.00']\n",
      "CCI43 ['CCI43/checkm_quality_assessment.txt:', 'CCI43_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '124', '1210', '21', '3', '1', '0', '86.41', '2.34', '0.00']\n",
      "CCI44 ['CCI44/checkm_quality_assessment.txt:', 'CCI44_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1205', '22', '4', '1', '0', '85.97', '2.70', '0.00']\n",
      "CCI45 ['CCI45/checkm_quality_assessment.txt:', 'CCI45_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '124', '1208', '24', '2', '1', '0', '86.46', '2.81', '0.00']\n",
      "CCI46 ['CCI46/checkm_quality_assessment.txt:', 'CCI46_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '5', '1349', '5', '0', '0', '0', '99.40', '0.47', '0.00']\n",
      "CCI47 ['CCI47/checkm_quality_assessment.txt:', 'CCI47_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1349', '8', '0', '0', '0', '99.70', '0.51', '0.00']\n",
      "CCI48 ['CCI48/checkm_quality_assessment.txt:', 'CCI48_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '124', '1210', '21', '3', '1', '0', '86.46', '2.53', '0.00']\n",
      "CCI49 ['CCI49/checkm_quality_assessment.txt:', 'CCI49_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1350', '8', '0', '0', '0', '99.99', '0.64', '12.50']\n",
      "CCI4 ['CCI4/checkm_quality_assessment.txt:', 'CCI4_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1355', '3', '0', '0', '0', '99.99', '0.38', '0.00']\n",
      "CCI50 ['CCI50/checkm_quality_assessment.txt:', 'CCI50_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '6', '1351', '2', '0', '0', '0', '99.42', '0.31', '0.00']\n",
      "CCI51 ['CCI51/checkm_quality_assessment.txt:', 'CCI51_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1349', '8', '0', '0', '0', '99.70', '0.51', '0.00']\n",
      "CCI52 ['CCI52/checkm_quality_assessment.txt:', 'CCI52_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1207', '21', '2', '2', '0', '86.06', '2.60', '0.00']\n",
      "CCI53 ['CCI53/checkm_quality_assessment.txt:', 'CCI53_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1207', '21', '2', '2', '0', '86.06', '2.60', '0.00']\n",
      "CCI54 ['CCI54/checkm_quality_assessment.txt:', 'CCI54_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '126', '1208', '21', '2', '2', '0', '86.12', '2.60', '0.00']\n",
      "CCI55 ['CCI55/checkm_quality_assessment.txt:', 'CCI55_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '131', '1207', '17', '2', '2', '0', '85.81', '2.41', '0.00']\n",
      "CCI56 ['CCI56/checkm_quality_assessment.txt:', 'CCI56_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '123', '1211', '21', '3', '1', '0', '86.55', '2.79', '0.00']\n",
      "CCI57 ['CCI57/checkm_quality_assessment.txt:', 'CCI57_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '124', '1210', '21', '3', '1', '0', '86.41', '2.34', '0.00']\n",
      "CCI58 ['CCI58/checkm_quality_assessment.txt:', 'CCI58_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1352', '6', '0', '0', '0', '99.99', '0.95', '0.00']\n",
      "CCI59 ['CCI59/checkm_quality_assessment.txt:', 'CCI59_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1349', '9', '0', '0', '0', '99.99', '0.55', '0.00']\n",
      "CCI5 ['CCI5/checkm_quality_assessment.txt:', 'CCI5_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '97', '1225', '35', '2', '0', '0', '88.02', '3.35', '0.00']\n",
      "CCI60 ['CCI60/checkm_quality_assessment.txt:', 'CCI60_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1351', '7', '0', '0', '0', '99.99', '1.00', '0.00']\n",
      "CCI61 ['CCI61/checkm_quality_assessment.txt:', 'CCI61_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1349', '8', '0', '0', '0', '99.70', '0.51', '0.00']\n",
      "CCI62 ['CCI62/checkm_quality_assessment.txt:', 'CCI62_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1359', '0', '0', '0', '0', '0', '0.00', '0.00', '0.00']\n",
      "CCI63 ['CCI63/checkm_quality_assessment.txt:', 'CCI63_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '127', '1208', '20', '2', '2', '0', '86.06', '2.50', '0.00']\n",
      "CCI64 ['CCI64/checkm_quality_assessment.txt:', 'CCI64_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1355', '3', '0', '0', '0', '99.99', '0.32', '0.00']\n",
      "CCI65 ['CCI65/checkm_quality_assessment.txt:', 'CCI65_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '88', '1236', '34', '1', '0', '0', '87.88', '3.07', '8.11']\n",
      "CCI66 ['CCI66/checkm_quality_assessment.txt:', 'CCI66_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '90', '1231', '37', '1', '0', '0', '87.77', '2.80', '0.00']\n",
      "CCI67 ['CCI67/checkm_quality_assessment.txt:', 'CCI67_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '93', '1250', '15', '1', '0', '0', '88.90', '1.33', '0.00']\n",
      "CCI68 ['CCI68/checkm_quality_assessment.txt:', 'CCI68_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1348', '11', '0', '0', '0', '0', '0.82', '0.00', '0.00']\n",
      "CCI69 ['CCI69/checkm_quality_assessment.txt:', 'CCI69_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '3', '97', '1221', '36', '2', '0', '99.98', '90.20', '47.13']\n",
      "CCI6 ['CCI6/checkm_quality_assessment.txt:', 'CCI6_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1356', '2', '0', '0', '0', '99.99', '0.31', '0.00']\n",
      "CCI70 ['CCI70/checkm_quality_assessment.txt:', 'CCI70_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '3', '82', '69', '1157', '33', '15', '99.98', '182.63', '55.16']\n",
      "CCI71 ['CCI71/checkm_quality_assessment.txt:', 'CCI71_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '1353', '5', '0', '0', '0', '99.95', '0.40', '0.00']\n",
      "CCI72 ['CCI72/checkm_quality_assessment.txt:', 'CCI72_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '218', '947', '186', '7', '1', '0', '74.77', '13.32', '14.08']\n",
      "CCI73 ['CCI73/checkm_quality_assessment.txt:', 'CCI73_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '124', '1107', '97', '28', '3', '0', '85.53', '13.83', '47.24']\n",
      "CCI74 ['CCI74/checkm_quality_assessment.txt:', 'CCI74_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '1', '69', '1274', '14', '1', '0', '99.99', '91.05', '52.34']\n",
      "CCI75 ['CCI75/checkm_quality_assessment.txt:', 'CCI75_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '103', '1219', '36', '0', '0', '1', '88.16', '3.07', '0.00']\n",
      "CCI76 ['CCI76/checkm_quality_assessment.txt:', 'CCI76_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '125', '1149', '62', '23', '0', '0', '85.03', '8.49', '29.77']\n",
      "CCI77 ['CCI77/checkm_quality_assessment.txt:', 'CCI77_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '404', '181', '95', '565', '69', '45', '61.63', '95.67', '2.94']\n",
      "CCI78 ['CCI78/checkm_quality_assessment.txt:', 'CCI78_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '68', '1276', '14', '1', '0', '0', '90.17', '0.68', '5.88']\n",
      "CCI79 ['CCI79/checkm_quality_assessment.txt:', 'CCI79_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1349', '8', '0', '0', '0', '99.70', '0.90', '25.00']\n",
      "CCI7 ['CCI7/checkm_quality_assessment.txt:', 'CCI7_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '97', '1227', '34', '1', '0', '0', '87.59', '2.43', '5.41']\n",
      "CCI80 ['CCI80/checkm_quality_assessment.txt:', 'CCI80_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '94', '1228', '34', '3', '0', '0', '87.88', '2.98', '4.65']\n",
      "CCI8 ['CCI8/checkm_quality_assessment.txt:', 'CCI8_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1347', '10', '0', '0', '0', '99.70', '1.26', '20.00']\n",
      "CCI9 ['CCI9/checkm_quality_assessment.txt:', 'CCI9_contigs', 'Klebsiella', '(5)', '29', '1359', '348', '2', '1347', '10', '0', '0', '0', '99.70', '1.26', '20.00']\n"
     ]
    }
   ],
   "source": [
    "checkm_location = \"/Users/dmatute/Documents/CAMRA/Bioinformatics/IFAIN/CAMRA_BatchA_121022_dir/QC/batch1_combined_checkM_quality_assessment.txt\"\n",
    "batch_name = \"CAMRA_BatchA_121022_dir\"\n",
    "\n",
    "#Make directories from checkm files that derrick made\n",
    "if os.path.isfile(checkm_location):\n",
    "    checkm_file = open(checkm_location, \"r\")\n",
    "    content=checkm_file.readlines()\n",
    "    for i in range (0,len(content)):\n",
    "        #print(content[i][0:3].strip(), content[i])\n",
    "        if content[i][0:3].strip() == \"CCI\":\n",
    "            checkm_metrics = content[i].split()\n",
    "            sample_name = checkm_metrics[0].split(\"/\")[0]\n",
    "            os.mkdir(f\"/Users/dmatute/Documents/CAMRA/Bioinformatics/IFAIN/CAMRA_BatchA_121022_dir/Assembly/{sample_name}\")\n",
    "            # checkm_markerset = checkm_metrics[1]\n",
    "            # checkm_completness = checkm_metrics[12]\n",
    "            # checkm_contamination = checkm_metrics[13]\n",
    "            # checkm_heterogeneity = checkm_metrics[14]\n",
    "            print(sample_name, checkm_metrics)\n",
    "        # else:\n",
    "        #     message = message +\"    CheckM unavailable\"\n",
    "    checkm_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "241d4a6c-0747-42da-a79f-7dad9ec1f131",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "list"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = [1,2,3]\n",
    "type(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b8a70c0-069e-4f50-946f-91ce0fe4163e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
